### PR TITLE
switch all templates to a generic one

### PR DESCRIPTION
**Pull request by avdev4j**
As we discuss on #25 @murdos here is a way we can do to use only one template for all blueprint generated.

Also we could complete unit tests to assert that the file is created with the good values.

### CODE REVIEW

**Comment by avdev4j** [comment]
I added common generator and tests linked to.
This PR can replace both #24 #25 ;)

File: generators/app/index.js [hunk 1]
```diff
@@ -61,6 +73,10 @@ module.exports = class extends Generator {
                         name: 'server',
                         value: 'server'
                     },
+                    {
```
Comment (murdos) [inline generators/app/index.js]: I think the whole `choices` array can be derived/constructed from the keys of `this.generators` that you've defined above.

File: generators/app/templates/generators/_index.js.ejs [hunk 2]
```diff
@@ -76,8 +92,20 @@ module.exports = class extends LanguagesGenerator {
     }
 
     get writing() {
```
Comment (murdos) [inline generators/app/templates/generators/_index.js.ejs]: I think you should fix #19 at the same time, by adding all methods that can exist in a blueprint generator:
https://github.com/jhipster/generator-jhipster/blob/master/generators/generator-base-blueprint.js

I think only `install` is missing.

File: generators/app/index.js [hunk 3]
```diff
@@ -195,45 +212,16 @@ module.exports = class extends Generator {
...
-        }
-        if (this.blueprintSubs.includes('spring-controller')) {
-            // copy files for the spring-controller generator
-            this.template('generators/spring-controller/_index.js', 'generators/spring-controller/index.js');
-        }
-        if (this.blueprintSubs.includes('spring-service')) {
-            // copy files for the spring-service generator
-            this.template('generators/spring-service/_index.js', 'generators/spring-service/index.js');
-        }
+        this.blueprintSubs.forEach(generator => {
```
Comment (murdos) [inline generators/app/index.js]: You can include in this loop the handling of `test/_subgen.spec.ejs` and the `.yo-rc.json` : 
```
if (!generator.startsWith('entity-')) {
    this.template('test/_subgen.spec.ejs', `test/${generator}.spec.js`);
}
if (['entity', 'languages', 'spring-controller', 'spring-service'].includes(generator))) {
    this.template('test/templates/ngx-blueprint/.yo-rc.json.ejs', 'test/templates/ngx-blueprint/.yo-rc.json');
}
```

File: generators/app/index.js (hunk 1 above)
Comment (avdev4j) [reply generators/app/index.js]: done ;)

File: generators/app/index.js (hunk 3 above)
Comment (avdev4j) [reply generators/app/index.js]: done ;)

**Comment by avdev4j** [comment]
I probably found an issue : 
at this [line](https://github.com/jhipster/generator-jhipster-blueprint/blob/0c9f8d1069a8040d8af4a787dd84e21572c1d582/generators/app/templates/generators/_index.js.ejs#L20) we should handle other generators than server.

About this, is this mandatory ? We already call this method in parent constructor.

**Comment by avdev4j** [comment]
I added the missing part for sub Generators but I wonder if it complete. Could you please validate this? (especially for the common generator).

```        <% if (subGenerator === 'entity-server') { %>
        if (jhContext.databaseType === 'cassandra') {
            this.pkType = 'UUID';
        }
        <% } %>
        // This sets up options for this sub generator and is being reused from JHipster
        <% if(subGenerator === 'server') { %>
        jhContext.setupServerOptions(this, jhContext);
        <% } else if(subGenerator === 'client') { %>
        jhContext.setupClientOptions(this, jhContext);
        <% } else if (subGenerator === 'entity') { %>
        jhContext.setupEntityOptions(this, jhContext, this);
        <% } else if (subGenerator === 'common') {%>
        jhContext.setupServerOptions(this, jhContext);
        jhContext.setupClientOptions(this, jhContext);
        <% } %>```

**Comment by murdos** [comment]
> 
> About this, is this mandatory ? We already call this method in parent constructor.

I have no clue... :) Let's just keep it for now in doubt :)



> 
> 
> I added the missing part for sub Generators but I wonder if it complete. Could you please validate this? (especially for the common generator).
> 
> ```
>         if (jhContext.databaseType === 'cassandra') {
>             this.pkType = 'UUID';
>         }
>         <% } %>
>         // This sets up options for this sub generator and is being reused from JHipster
>         <% if(subGenerator === 'server') { %>
>         jhContext.setupServerOptions(this, jhContext);
>         <% } else if(subGenerator === 'client') { %>
>         jhContext.setupClientOptions(this, jhContext);
>         <% } else if (subGenerator === 'entity') { %>
>         jhContext.setupEntityOptions(this, jhContext, this);
>         <% } else if (subGenerator === 'common') {%>
>         jhContext.setupServerOptions(this, jhContext);
>         jhContext.setupClientOptions(this, jhContext);
>         <% } %>```
> ```

I'm fine with that (including common), I just have a minor hassle: the comment `// This sets up options for this sub generator and is being reused from JHipster` is always here whereas it's only needed for server, client, entity and common.

Could you fix it?
Then I'll probably merge this PR. I've done some tests locally, there are only minor differences between projects generated before and after your changes.

Good job, thanks!!

**Comment by avdev4j** [comment]
Oh yes. Finally I added the comment in all "if" blocks. Probalbly I could avoiding some duplicated here but I think is clearer like that. 

If there is minor differences, I could fix theml if you want, we have time to do a clean template ;).

**Comment by murdos** [comment]
> 
> If there is minor differences, I could fix theml if you want, we have time to do a clean template ;).

No, it's ok. For example I've seen added empty new line, but I think it's better that way :)

@pascalgrimaud The integration tests are failing, and it's not related to the PR. Could you help on the subject? The error is below, and I think it's related to the npm linking we're doing...
```
[INFO] --- frontend-maven-plugin:1.6:npm (npm install) @ jhipster-blueprint-application ---
[INFO] Running 'npm install' in /home/travis/app
[WARNING] npm WARN acorn-dynamic-import@4.0.0 requires a peer of acorn@^6.0.0 but none is installed. You must install peer dependencies yourself.
[WARNING] npm WARN bootstrap@4.3.1 requires a peer of jquery@1.9.1 - 3 but none is installed. You must install peer dependencies yourself.
[WARNING] npm WARN bootstrap@4.3.1 requires a peer of popper.js@^1.14.7 but none is installed. You must install peer dependencies yourself.
[WARNING] npm WARN jest-preset-angular@6.0.2 requires a peer of jest@^23.0.0 but none is installed. You must install peer dependencies yourself.
[WARNING] npm WARN ts-jest@23.1.4 requires a peer of babel-core@^6.0.0 || ^7.0.0-0 but none is installed. You must install peer dependencies yourself.
[ERROR] 
[ERROR] npm ERR! path /home/travis/app/node_modules/generator-jhipster-helloworld/node_modules/.bin/jhipster
[ERROR] npm ERR! code EEXIST
[ERROR] npm ERR! Refusing to delete /home/travis/app/node_modules/generator-jhipster-helloworld/node_modules/.bin/jhipster: containing path /home/travis/app/node_modules/generator-jhipster-helloworld/node_modules/generator-jhipster isn't under npm's control
[ERROR] npm ERR! File exists: /home/travis/app/node_modules/generator-jhipster-helloworld/node_modules/.bin/jhipster
[ERROR] npm ERR! Move it away, and try again.
[ERROR] 
[ERROR] npm ERR! A complete log of this run can be found in:
[ERROR] npm ERR!     /home/travis/.npm/_logs/2019-03-08T07_41_59_515Z-debug.log
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

**Comment by pascalgrimaud** [comment]
Sorry not the time right now... but self assigned this PR, so I can review later

**Comment by pascalgrimaud** [comment]
@avdev4j : I'm starting to review and work on it, to understand why Travis failed here.

**Pull request merged**